File: mac/Grimly/ViewModels/EditorPopupViewModel.swift
import SwiftUI
import AppKit
import Combine

enum ConnectionStatus {
    case checking
    case connected
    case modelNotLoaded
    case foundryNotRunning
    case foundryNotInstalled

    var label: String {
        switch self {
        case .checking: return "Checking local LLM"
        case .connected: return "Connected to local LLM"
        case .modelNotLoaded: return "Not connected to local LLM (model not loaded)"
        case .foundryNotRunning: return "Not connected to local LLM (Foundry not running)"
        case .foundryNotInstalled: return "Not connected to local LLM (Foundry not installed)"
        }
    }

    var color: Color {
        switch self {
        case .checking: return Color(white: 0.5)
        case .connected: return Color(red: 0.31, green: 0.82, blue: 0.35)
        case .modelNotLoaded: return Color(red: 0.9, green: 0.7, blue: 0.2)
        case .foundryNotRunning: return Color(red: 0.86, green: 0.31, blue: 0.31)
        case .foundryNotInstalled: return Color(red: 0.63, green: 0.2, blue: 0.2)
        }
    }
}

@MainActor
class EditorPopupViewModel: ObservableObject {
    private let foundryClient: FoundryLocalClient
    private let foundryManager: FoundryManager
    private let clipboardService: ClipboardService
    private let diffService: TextDiffService
    private let connectionMonitor: ConnectionMonitor
    private let properNouns = ProperNounService()
    private let spellChecker = SpellCheckerService()
    private lazy var codeChecker = GrammarChecker(spellChecker: spellChecker, properNouns: properNouns)
    private let readabilityService = ReadabilityService()
    private let apStyleCodePass = APStyleCodePass()
    private lazy var apStylePipeline = APStylePipeline(codePass: apStyleCodePass, client: foundryClient)
    private var currentTask: Task<Void, Never>?
    private var undoStack: [String] = []
    private var preRevisionText: String = ""
    private var cancellables = Set<AnyCancellable>()

    @Published var workingText: String = ""
    @Published var reviewSegments: [ReviewSegment] = []
    @Published var selectedMode: EditingMode = .fixGrammar
    @Published var customPrompt: String = ""
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    @Published var hasResult: Bool = false
    @Published var canUndo: Bool = false
    @Published var isReviewing: Bool = false
    @Published var appliedModes: Set<EditingMode> = []
    @Published var connectionStatus: ConnectionStatus = .checking
    /// True while the app-level ConnectionMonitor is actively retrying.
    /// The View reads this to show a subtler "Reconnecting…" banner
    /// instead of a red error when the LLM is momentarily unavailable —
    /// the retry will usually recover before the user notices.
    @Published var isReconnecting: Bool = false
    @Published var readabilityScore: Double = 0
    @Published var readabilityLabel: String = ""
    @Published var wordCount: Int = 0
    @Published var charCount: Int = 0

    /// Live deterministic violations (grammar, spelling, punctuation, …).
    /// Re-populated ~400 ms after `workingText` last changed.
    @Published var violations: [Violation] = []
    var hasViolations: Bool { !violations.isEmpty }
    /// True when at least one violation has a deterministic auto-fix —
    /// drives Quick Fix button visibility.
    var hasAutoFixableViolations: Bool { violations.contains(where: { $0.canAutoFix }) }

    /// Hint text shown above the violations list.
    let quickFixHint = "Click Quick Fix for the mechanical corrections. Use Fix Grammar for AI-assisted revisions of the rest."

    var previousApp: NSRunningApplication?
    var onRequestClose: (() -> Void)?
    var onReviewSegmentsChanged: (() -> Void)?

    var isCustomMode: Bool { selectedMode == .customPrompt }

    init(
        foundryClient: FoundryLocalClient,
        foundryManager: FoundryManager,
        clipboardService: ClipboardService,
        diffService: TextDiffService,
        connectionMonitor: ConnectionMonitor
    ) {
        self.foundryClient = foundryClient
        self.foundryManager = foundryManager
        self.clipboardService = clipboardService
        self.diffService = diffService
        self.connectionMonitor = connectionMonitor

        // Seed from the monitor's current state so the popup renders with
        // the right status the moment it opens (no "checking" flash while
        // Combine settles).
        self.connectionStatus = connectionMonitor.status
        self.isReconnecting = connectionMonitor.isReconnecting

        // Mirror the app-level monitor's state into VM-published fields so
        // the SwiftUI view can observe them without touching the monitor
        // directly. All reconnect logic lives in ConnectionMonitor — the
        // VM used to run its own reconnect loop, but that only worked
        // while the popup was open. The monitor keeps recovering even when
        // the user has closed the window.
        connectionMonitor.$status
            .receive(on: RunLoop.main)
            .sink { [weak self] s in
                self?.connectionStatus = s
                // Clear the red error banner once we're back on the air.
                if s == .connected, let self, self.errorMessage?.starts(with: "Cannot connect") == true {
                    self.errorMessage = nil
                }
            }
            .store(in: &cancellables)

        connectionMonitor.$isReconnecting
            .receive(on: RunLoop.main)
            .sink { [weak self] r in self?.isReconnecting = r }
            .store(in: &cancellables)

        // Update readability score whenever working text changes
        $workingText
            .debounce(for: .milliseconds(200), scheduler: RunLoop.main)
            .sink { [weak self] _ in self?.updateReadability() }
            .store(in: &cancellables)

        // Live deterministic grammar / spelling / punctuation check —
        // independent debounce so the panel updates ~400 ms after the user
        // stops typing. Mirrors the Windows VM. The checker runs on the
        // main actor; for our text volumes (a few thousand words at most)
        // the regex sweep is well under a frame budget.
        $workingText
            .debounce(for: .milliseconds(400), scheduler: RunLoop.main)
            .sink { [weak self] text in self?.runLiveCheck(on: text) }
            .store(in: &cancellables)
    }

    private func runLiveCheck(on text: String) {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            violations = []
            return
        }
        violations = codeChecker.check(text)
    }

    /// Re-case the working text and show the result as a reviewable diff.
    /// Deterministic — no LLM round-trip. Used by the "Case" dropdown for
    /// AP title case and Chicago title case.
    func applyCase(_ style: CaseStyle) {
        let rewritten = CaseFormatter.apply(workingText, style: style)
        guard rewritten != workingText else { return }

        preRevisionText = workingText
        let diffs = diffService.computeDiff(original: preRevisionText, corrected: rewritten)
        let segments = diffService.groupIntoSegments(diffs)
        reviewSegments = segments

        if segments.contains(where: { $0.isChange }) {
            undoStack.append(preRevisionText)
            canUndo = true
            isReviewing = true
            hasResult = true
            rebuildWorkingText()
            onReviewSegmentsChanged?()
        }
    }

    /// AP Style — run the two-pass pipeline (deterministic code pass +
    /// narrow LLM pass) on the working text and show the result as a
    /// reviewable diff. The code pass alone is enough for most changes;
    /// the LLM pass adds attribution-verb + passive-voice + editorial-
    /// framing rewrites. Prompt-echo guard in the pipeline protects the
    /// diff from small-model prompt leaks.
    func runApStyle() {
        currentTask?.cancel()
        let task = Task {
            isLoading = true
            errorMessage = nil
            isReviewing = false

            preRevisionText = workingText

            let result = await apStylePipeline.run(preRevisionText)
            if Task.isCancelled { isLoading = false; return }

            let diffs = diffService.computeDiff(original: preRevisionText, corrected: result)
            let segments = diffService.groupIntoSegments(diffs)
            reviewSegments = segments

            let hasChanges = segments.contains { $0.isChange }
            if hasChanges {
                undoStack.append(preRevisionText)
                canUndo = true
                isReviewing = true
                hasResult = true
                rebuildWorkingText()
                onReviewSegmentsChanged?()
            } else {
                errorMessage = "No AP Style changes suggested."
            }
            isLoading = false
        }
        currentTask = task
    }

    /// Quick Fix — apply every deterministic fix the live checker found, as
    /// a single reviewable diff. Mirrors the accept/reject UX used by Fix
    /// Grammar so the user reviews before committing rather than seeing
    /// changes applied silently.
    func applyQuickFixes() {
        let fixed = codeChecker.applyAutoFixes(workingText)
        guard fixed != workingText else { return }

        preRevisionText = workingText
        let diffs = diffService.computeDiff(original: preRevisionText, corrected: fixed)
        let segments = diffService.groupIntoSegments(diffs)
        reviewSegments = segments

        if segments.contains(where: { $0.isChange }) {
            undoStack.append(preRevisionText)
            canUndo = true
            isReviewing = true
            hasResult = true
            rebuildWorkingText()
            onReviewSegmentsChanged?()
        }
    }

    /// Trigger an immediate connection check on the shared monitor. Used
    /// for the user tapping the status LED and for post-error nudges.
    func refreshConnectionStatus() {
        Task {
            await connectionMonitor.refresh()
        }
    }

    func isModeApplied(_ mode: EditingMode) -> Bool {
        appliedModes.contains(mode)
    }

    func setCapturedText(_ text: String) {
        workingText = text
        updateReadability()
    }

    private func updateReadability() {
        guard !workingText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            readabilityScore = 0
            readabilityLabel = ""
            wordCount = 0
            charCount = 0
            return
        }

        wordCount = workingText
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .count
        charCount = workingText.count
        readabilityScore = readabilityService.calculateFleschReadingEase(workingText)

        let wordLabel = wordCount == 1 ? "word" : "words"
        let charLabel = charCount == 1 ? "char" : "chars"
        readabilityLabel = "\(wordCount.formatted()) \(wordLabel) · \(charCount.formatted()) \(charLabel) · Readability \(String(format: "%.1f", readabilityScore))"
    }

    func process() {
        currentTask?.cancel()

        let task = Task {
            isLoading = true
            errorMessage = nil
            isReviewing = false

            do {
                preRevisionText = workingText

                let result = try await foundryClient.getEditedText(
                    originalText: preRevisionText,
                    mode: selectedMode,
                    customPrompt: selectedMode == .customPrompt ? customPrompt : nil
                )

                if Task.isCancelled { return }

                let diffs = diffService.computeDiff(original: preRevisionText, corrected: result)
                let segments = diffService.groupIntoSegments(diffs)

                reviewSegments = segments

                appliedModes.insert(selectedMode)

                let hasChanges = segments.contains { $0.isChange }
                if hasChanges {
                    errorMessage = nil
                    undoStack.append(preRevisionText)
                    canUndo = true
                    isReviewing = true
                    hasResult = true
                    rebuildWorkingText()
                    onReviewSegmentsChanged?()
                } else {
                    errorMessage = "No changes suggested."
                }
            } catch is CancellationError {
                // Cancelled, no action needed
            } catch is URLError {
                // Client already retried once with a refreshed endpoint;
                // both attempts failed. Poke the monitor so its reconnect
                // ladder starts (or restarts) immediately, and use a
                // gentler banner if a reconnect is already in flight —
                // no need to shout when the app is already recovering.
                if isReconnecting {
                    errorMessage = "Reconnecting to local LLM — try again shortly."
                } else {
                    errorMessage = "Cannot connect to Foundry Local. Is it running?"
                }
                refreshConnectionStatus()
            } catch {
                errorMessage = "Error: \(error.localizedDescription)"
                refreshConnectionStatus()
            }

            isLoading = false
        }

        currentTask = task
    }

    func toggleChange(_ segmentId: Int) {
        guard let segment = reviewSegments.first(where: { $0.id == segmentId && $0.isChange }) else { return }
        segment.toggle()
        rebuildWorkingText()
        onReviewSegmentsChanged?()
    }

    func setChangeStates(_ segmentIds: [Int], state: ChangeState) {
        for id in segmentIds {
            if let seg = reviewSegments.first(where: { $0.id == id && $0.isChange }) {
                seg.state = state
            }
        }
        rebuildWorkingText()
        onReviewSegmentsChanged?()
    }

    func acceptAllChanges() {
        for seg in reviewSegments where seg.isChange {
            seg.state = .accepted
        }
        rebuildWorkingText()
        onReviewSegmentsChanged?()
    }

    func rejectAllChanges() {
        for seg in reviewSegments where seg.isChange {
            seg.state = .rejected
        }
        rebuildWorkingText()
        onReviewSegmentsChanged?()
    }

    func applyReview() {
        isReviewing = false
        reviewSegments = []
    }

    private func rebuildWorkingText() {
        workingText = reviewSegments.map(\.resolvedText).joined()
    }

    func undo() {
        guard !undoStack.isEmpty else { return }
        workingText = undoStack.removeLast()
        canUndo = !undoStack.isEmpty
        isReviewing = false
        hasResult = !undoStack.isEmpty
        reviewSegments = []
    }

    func accept() {
        if isReviewing {
            applyReview()
        }

        let textToPaste = workingText
        onRequestClose?()

        Task {
            try? await Task.sleep(nanoseconds: 100_000_000)
            await clipboardService.pasteText(textToPaste, previousApp: previousApp)
        }
    }

    func dismiss() {
        currentTask?.cancel()
        onRequestClose?()
    }

    func copyResult() {
        guard !workingText.isEmpty else { return }
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(workingText, forType: .string)
    }
}

File: src/Grimly.Core/ViewModels/EditorPopupViewModel.cs
using System.Collections.ObjectModel;
using System.Net.Http;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using Grimly.Hosting;
using Grimly.Models;
using Grimly.Services;

namespace Grimly.ViewModels;

public partial class EditorPopupViewModel : ObservableObject
{
    private readonly IFoundryLocalClient _foundryClient;
    private readonly IClipboardService _clipboardService;
    private readonly ITextDiffService _diffService;
    private readonly IFoundryManager _foundryManager;
    private readonly IReadabilityService _readabilityService;
    private readonly IWordReadabilityService? _wordReadabilityService;
    private readonly IGrammarChecker _codeChecker;
    private readonly ISettingsService _settingsService;
    private readonly BrandingOptions _branding;
    private CancellationTokenSource? _cts;
    private CancellationTokenSource? _liveCheckCts;
    private readonly Stack<string> _undoStack = new();
    private string _preRevisionText = "";
    private readonly HashSet<EditingMode> _appliedModes = new();
    private int _consecutiveTimeouts;

    public event Action? AppliedModesChanged;

    [ObservableProperty]
    private string _workingText = "";

    [ObservableProperty]
    private ObservableCollection<ReviewSegment> _reviewSegments = [];

    [ObservableProperty]
    private EditingMode _selectedMode;

    [ObservableProperty]
    private string _customPrompt = "";

    [ObservableProperty]
    private bool _isLoading;

    [ObservableProperty]
    private string? _errorMessage;

    [ObservableProperty]
    private bool _hasResult;

    [ObservableProperty]
    private bool _canUndo;

    [ObservableProperty]
    private bool _isReviewing; // true when showing accept/reject UI

    [ObservableProperty]
    private bool _showRetry;

    [ObservableProperty]
    private ConnectionStatus _connectionStatus = ConnectionStatus.Unknown;

    [ObservableProperty]
    private string _connectionStatusText = "Checking local LLM...";

    [ObservableProperty]
    private double _readabilityScore;

    [ObservableProperty]
    private string _readabilityLabel = "";

    [ObservableProperty]
    private int _wordCount;

    [ObservableProperty]
    private int _charCount;

    // ===== Match Word state =====
    /// <summary>
    /// True when Microsoft Word is installed on this machine and can be called
    /// via COM to compute the precise Flesch score. Drives visibility of the
    /// "Match Word" button.
    /// </summary>
    [ObservableProperty]
    private bool _isWordMatchAvailable;

    /// <summary>True while the Match Word call is in flight.</summary>
    [ObservableProperty]
    private bool _isMatchingWord;

    /// <summary>
    /// True when the displayed readability score came from Word (precise) rather
    /// than the local heuristic (approximate). Any text change flips this back
    /// to false.
    /// </summary>
    [ObservableProperty]
    private bool _isWordMatched;

    /// <summary>
    /// True when the popup opened as a manual-paste fallback — the host
    /// triggered the popup (hotkey or floating icon), but UIA couldn't read
    /// a selection from the target app (Electron, custom text surfaces,
    /// accessibility denied). The UI shows a specific hint telling the user
    /// to paste their text into the working-text box.
    /// </summary>
    [ObservableProperty]
    private bool _isManualPasteMode;

    // ===== Live grammar-check state =====
    /// <summary>Deterministic violations from the live grammar checker.</summary>
    public ObservableCollection<Violation> Violations { get; }

    /// <summary>Convenience flag for XAML visibility. Tracks <c>Violations.Count &gt; 0</c>.</summary>
    public bool HasViolations => Violations.Count > 0;

    /// <summary>Hint text shown above the violations list.</summary>
    public string QuickFixHint =>
        "Click Quick Fix for the mechanical corrections. Use Fix Grammar for AI-assisted revisions of the rest.";

    public bool IsCustomMode => SelectedMode == EditingMode.CustomPrompt;

    partial void OnSelectedModeChanged(EditingMode value)
    {
        OnPropertyChanged(nameof(IsCustomMode));
    }

    public IReadOnlyList<EditingMode> AvailableModes { get; } = Enum.GetValues<EditingMode>();

    private static readonly char[] WordSeparators = { ' ', '\t', '\n', '\r', '\u00A0' };

    public event Action? RequestClose;
    public event Action? ReviewSegmentsChanged;
    public IntPtr PreviousForegroundWindow { get; set; }

    public EditorPopupViewModel(
        IFoundryLocalClient foundryClient,
        IClipboardService clipboardService,
        ITextDiffService diffService,
        IFoundryManager foundryManager,
        IReadabilityService readabilityService,
        IGrammarChecker codeChecker,
        ISettingsService settingsService,
        BrandingOptions branding,
        IWordReadabilityService? wordReadabilityService = null)
    {
        _foundryClient = foundryClient;
        _clipboardService = clipboardService;
        _diffService = diffService;
        _foundryManager = foundryManager;
        _readabilityService = readabilityService;
        _wordReadabilityService = wordReadabilityService;
        _codeChecker = codeChecker;
        _settingsService = settingsService;
        _branding = branding;

        IsWordMatchAvailable = wordReadabilityService?.IsAvailable ?? false;

        Violations = [];
        Violations.CollectionChanged += (_, _) =>
        {
            OnPropertyChanged(nameof(HasViolations));
            OnPropertyChanged(nameof(HasAutoFixableViolations));
            ApplyQuickFixesCommand.NotifyCanExecuteChanged();
        };
    }

    partial void OnWorkingTextChanged(string value)
    {
        UpdateReadability(value);
        // Once the user has pasted anything non-empty into the internal
        // buffer, we're no longer in the "can't read selection" state —
        // revert to the normal hint wording.
        if (!string.IsNullOrEmpty(value) && IsManualPasteMode)
            IsManualPasteMode = false;

        ScheduleLiveCheck(value);
    }

    /// <summary>
    /// Debounced live deterministic style/grammar/spelling check. Runs the
    /// in-process <see cref="IGrammarChecker"/> ~400 ms after the user
    /// stops typing so the violations panel reflects the current text without
    /// requiring a button click.
    /// </summary>
    private void ScheduleLiveCheck(string snapshot)
    {
        _liveCheckCts?.Cancel();
        _liveCheckCts = new CancellationTokenSource();
        var ct = _liveCheckCts.Token;

        _ = Task.Run(async () =>
        {
            try
            {
                await Task.Delay(400, ct);
                if (ct.IsCancellationRequested) return;

                // Empty input = no violations, no work to do.
                IReadOnlyList<Violation> result = string.IsNullOrWhiteSpace(snapshot)
                    ? Array.Empty<Violation>()
                    : _codeChecker.Check(snapshot);

                if (ct.IsCancellationRequested) return;

                // Marshal back to the UI thread — Violations is bound to the
                // view, and ObservableCollection mutations must happen there.
                var dispatcher = System.Windows.Application.Current?.Dispatcher;
                if (dispatcher == null) return;
                await dispatcher.InvokeAsync(() =>
                {
                    if (ct.IsCancellationRequested) return;
                    Violations.Clear();
                    foreach (var v in result) Violations.Add(v);
                });
            }
            catch (OperationCanceledException) { }
            catch { /* live check failures are non-fatal */ }
        }, ct);
    }

    public async Task CheckConnectionAsync()
    {
        ConnectionStatus = ConnectionStatus.Unknown;
        ConnectionStatusText = "Checking...";

        var status = await _foundryManager.CheckConnectionAsync();
        ConnectionStatus = status;

        // Pull the active model name so the status line shows which LLM is
        // actually serving requests (helpful when the user has multiple models
        // or when the app falls back to a backup).
        var modelName = _settingsService.Load().ModelName;

        ConnectionStatusText = status switch
        {
            ConnectionStatus.Connected => $"Connected · {modelName}",
            ConnectionStatus.ServiceNotRunning => "Not connected to local LLM (service not running)",
            ConnectionStatus.ModelNotLoaded => "Not connected to local LLM (model not loaded)",
            ConnectionStatus.NotInstalled => "Not connected to local LLM (Foundry not installed)",
            ConnectionStatus.Error => "Local LLM not responding",
            _ => "Checking local LLM..."
        };

        // If the check confirms we're connected but an earlier request had set
        // a "cannot connect" error, reconcile the two — the connection is
        // actually fine, so downgrade the banner to a transient retry prompt
        // instead of showing a contradictory "Cannot connect" message.
        if (status == ConnectionStatus.Connected &&
            ErrorMessage == "Cannot connect to local LLM.")
        {
            ErrorMessage = "That request failed. Try again.";
        }
    }

    [RelayCommand]
    private async Task RetryConnectionAsync() => await CheckConnectionAsync();

    [RelayCommand]
    private async Task RetryAsync()
    {
        ShowRetry = false;
        ErrorMessage = null;
        await ProcessAsync();
    }

    private async Task FallbackToCpuModelAndRetryAsync()
    {
        ConnectionStatusText = "Switching to backup model...";
        ConnectionStatus = ConnectionStatus.Unknown;

        try
        {
            // Stop the service, restart, and load phi-4-mini (CPU, always reliable)
            var success = await _foundryManager.FallbackToCpuModelAsync();
            if (success)
            {
                _consecutiveTimeouts = 0;
                ConnectionStatus = ConnectionStatus.Connected;
                // Settings were just rewritten to the fallback model name.
                var backupName = _settingsService.Load().ModelName;
                ConnectionStatusText = $"Connected · {backupName} (backup)";
                ErrorMessage = null;
                await ProcessAsync();
            }
            else
            {
                ErrorMessage = "Could not switch models. Try restarting Foundry Local manually.";
                ShowRetry = true;
                _ = CheckConnectionAsync();
            }
        }
        catch
        {
            ErrorMessage = "Model switch failed.";
            ShowRetry = true;
        }
    }

    private async Task AutoReconnectAndRetryAsync()
    {
        ConnectionStatusText = "Restarting Foundry Local...";
        ConnectionStatus = ConnectionStatus.Unknown;

        var success = await _foundryManager.ForceReconnectAsync();
        if (success)
        {
            ConnectionStatus = ConnectionStatus.Connected;
            ConnectionStatusText = $"Connected · {_settingsService.Load().ModelName}";
            ErrorMessage = null;
            // Auto-retry
            await ProcessAsync();
        }
        else
        {
            ErrorMessage = "Auto-reconnect failed.";
            ShowRetry = true;
            _ = CheckConnectionAsync();
        }
    }

    [RelayCommand]
    private async Task ReconnectAsync()
    {
        ShowRetry = false;
        ErrorMessage = null;
        IsLoading = true;
        ConnectionStatusText = "Restarting Foundry Local...";
        ConnectionStatus = ConnectionStatus.Unknown;

        try
        {
            var success = await _foundryManager.ForceReconnectAsync();
            if (success)
            {
                _consecutiveTimeouts = 0;
                ConnectionStatus = ConnectionStatus.Connected;
                ConnectionStatusText = $"Connected · {_settingsService.Load().ModelName}";
                ErrorMessage = null;
                IsLoading = false;
                // Auto-retry the last command
                await ProcessAsync();
            }
            else
            {
                ErrorMessage = "Could not reconnect. Try restarting Foundry Local manually.";
                ShowRetry = true;
                _ = CheckConnectionAsync();
                IsLoading = false;
            }
        }
        catch
        {
            ErrorMessage = "Reconnection failed.";
            ShowRetry = true;
            IsLoading = false;
        }
    }

    private void UpdateReadability(string text)
    {
        if (string.IsNullOrWhiteSpace(text))
        {
            ReadabilityScore = 0;
            ReadabilityLabel = "";
            WordCount = 0;
            CharCount = 0;
            IsWordMatched = false;
            return;
        }

        // Word count: whitespace-delimited tokens, ignoring empty entries from
        // consecutive spaces. Char count: raw length including whitespace
        // (matches the Word / Google Docs default — users can mentally
        // subtract spaces if they want the no-spaces version).
        WordCount = text.Split(WordSeparators, StringSplitOptions.RemoveEmptyEntries).Length;
        CharCount = text.Length;

        ReadabilityScore = _readabilityService.CalculateFleschReadingEase(text);

        // Any text change invalidates a prior Word match — drop back to
        // the local estimate and re-render the label accordingly.
        IsWordMatched = false;
        RenderReadabilityLabel(fromWord: false);
        MatchWordCommand.NotifyCanExecuteChanged();
    }

    /// <summary>
    /// Rebuilds <see cref="ReadabilityLabel"/> from the current stats. The
    /// score is prefixed with "~" when it's the local estimate AND Word is
    /// available (signaling "click Match Word for the precise value"). When
    /// the score came from Word, it's shown plainly with a "(Word)" suffix.
    /// </summary>
    private void RenderReadabilityLabel(bool fromWord)
    {
        string scoreText;
        if (fromWord)
        {
            scoreText = $"Readability {ReadabilityScore:F1} (Word)";
        }
        else
        {
            // Only mark as estimate when Word is a realistic alternative.
            // On machines without Word there's nothing to compare against,
            // so the tilde would just be noise.
            var prefix = IsWordMatchAvailable ? "~" : "";
            scoreText = $"Readability {prefix}{ReadabilityScore:F1}";
        }

        ReadabilityLabel = $"{WordCount:N0} {(WordCount == 1 ? "word" : "words")} · " +
                           $"{CharCount:N0} {(CharCount == 1 ? "char" : "chars")} · " +
                           scoreText;
    }

    [RelayCommand(CanExecute = nameof(CanMatchWord))]
    private async Task MatchWordAsync()
    {
        if (_wordReadabilityService == null || !IsWordMatchAvailable) return;
        if (string.IsNullOrWhiteSpace(WorkingText)) return;

        IsMatchingWord = true;
        MatchWordCommand.NotifyCanExecuteChanged();
        try
        {
            var score = await _wordReadabilityService.CalculateFleschAsync(WorkingText);
            if (score.HasValue)
            {
                ReadabilityScore = score.Value;
                IsWordMatched = true;
                RenderReadabilityLabel(fromWord: true);
            }
            // null result = Word failed; leave local estimate visible. The
            // user can click again, or edit to retry. No error surfaced —
            // this is an enhancement, not a critical path.
        }
        finally
        {
            IsMatchingWord = false;
            MatchWordCommand.NotifyCanExecuteChanged();
        }
    }

    private bool CanMatchWord() =>
        IsWordMatchAvailable && !IsMatchingWord && !IsWordMatched &&
        !string.IsNullOrWhiteSpace(WorkingText);

    public bool IsModeApplied(EditingMode mode) => _appliedModes.Contains(mode);

    public void SetCapturedText(string text)
    {
        WorkingText = text;
    }

    [RelayCommand]
    private async Task ProcessAsync()
    {
        _cts?.Cancel();
        _cts = new CancellationTokenSource();
        var ct = _cts.Token;

        IsLoading = true;
        ErrorMessage = null;
        ShowRetry = false;
        IsReviewing = false;

        try
        {
            _preRevisionText = WorkingText;

            var result = await _foundryClient.GetEditedTextAsync(
                _preRevisionText, SelectedMode,
                SelectedMode == EditingMode.CustomPrompt ? CustomPrompt : null,
                ct);

            if (ct.IsCancellationRequested) return;

            // Compute diff and group into review segments
            var diffs = _diffService.ComputeDiff(_preRevisionText, result);
            var segments = _diffService.GroupIntoSegments(diffs);

            ReviewSegments = new ObservableCollection<ReviewSegment>(segments);

            // Check if there are any actual changes
            // Always mark the mode as applied
            _appliedModes.Add(SelectedMode);
            AppliedModesChanged?.Invoke();

            bool hasChanges = segments.Any(s => s.IsChange);
            if (hasChanges)
            {
                _undoStack.Push(_preRevisionText);
                _consecutiveTimeouts = 0; // success — reset timeout counter
                CanUndo = true;
                IsReviewing = true;
                HasResult = true;

                RebuildWorkingText();
                ReviewSegmentsChanged?.Invoke();
            }
            else
            {
                ErrorMessage = "No changes suggested.";
            }
        }
        catch (OperationCanceledException) { }
        catch (HttpRequestException ex)
        {
            var timedOut = ex.Message.Contains("timed out", StringComparison.OrdinalIgnoreCase);
            if (timedOut)
            {
                _consecutiveTimeouts++;
                if (_consecutiveTimeouts >= 2)
                {
                    // Second timeout — fall back to a CPU model
                    ErrorMessage = "Model not responding. Switching to a backup model...";
                    _ = FallbackToCpuModelAndRetryAsync();
                }
                else
                {
                    // First timeout — try reconnecting with the same model
                    ErrorMessage = "Request timed out. Reconnecting...";
                    _ = AutoReconnectAndRetryAsync();
                }
            }
            else
            {
                // Reset the status so the stale "Connected" dot doesn't linger
                // while CheckConnectionAsync re-evaluates in the background —
                // avoids showing "Connected" and "Cannot connect" at the same time.
                ConnectionStatus = ConnectionStatus.Unknown;
                ConnectionStatusText = "Checking...";
                ErrorMessage = "Cannot connect to local LLM.";
                ShowRetry = true;
                _ = CheckConnectionAsync();
            }
        }
        catch (Exception ex)
        {
            ErrorMessage = $"Error: {ex.Message}";
            ShowRetry = true;
        }
        finally
        {
            IsLoading = false;
        }
    }

    /// <summary>
    /// Quick Fix — apply every deterministic fix the live checker found, as
    /// a single reviewable diff. Mirrors the accept/reject UX used by Fix
    /// Grammar so the user reviews before committing, rather than seeing
    /// changes applied silently.
    ///
    /// Violations without a deterministic fix (ALL CAPS, absolute claims,
    /// spelling, etc.) stay in the panel — the hint text directs the user
    /// to the Fix Grammar button for those.
    /// </summary>
    [RelayCommand(CanExecute = nameof(CanApplyQuickFixes))]
    private void ApplyQuickFixes()
    {
        var fixed1 = _codeChecker.ApplyAutoFixes(WorkingText);
        if (fixed1 == WorkingText) return; // nothing to do

        _preRevisionText = WorkingText;
        var diffs = _diffService.ComputeDiff(_preRevisionText, fixed1);
        var segments = _diffService.GroupIntoSegments(diffs);
        ReviewSegments = new ObservableCollection<ReviewSegment>(segments);

        if (segments.Any(s => s.IsChange))
        {
            _undoStack.Push(_preRevisionText);
            CanUndo = true;
            IsReviewing = true;
            HasResult = true;
            RebuildWorkingText();
            ReviewSegmentsChanged?.Invoke();
        }
    }

    /// <summary>
    /// True when at least one auto-fixable violation is currently in the
    /// panel. Drives Quick Fix button enable state so users don't see a
    /// grayed-out button on text with only judgment-required violations.
    /// </summary>
    public bool HasAutoFixableViolations => Violations.Any(v => v.CanAutoFix);

    private bool CanApplyQuickFixes() => HasAutoFixableViolations && !IsLoading;

    public void ToggleChange(int segmentId)
    {
        var segment = ReviewSegments.FirstOrDefault(s => s.Id == segmentId && s.IsChange);
        if (segment == null) return;

        segment.Toggle();
        RebuildWorkingText();
        ReviewSegmentsChanged?.Invoke();
    }

    public void SetChangeStates(IEnumerable<int> segmentIds, ChangeState state)
    {
        foreach (var id in segmentIds)
        {
            var seg = ReviewSegments.FirstOrDefault(s => s.Id == id && s.IsChange);
            if (seg != null) seg.State = state;
        }
        RebuildWorkingText();
        ReviewSegmentsChanged?.Invoke();
    }

    [RelayCommand]
    private void AcceptAllChanges()
    {
        foreach (var seg in ReviewSegments.Where(s => s.IsChange))
            seg.State = ChangeState.Accepted;
        RebuildWorkingText();
        ReviewSegmentsChanged?.Invoke();
    }

    [RelayCommand]
    private void RejectAllChanges()
    {
        foreach (var seg in ReviewSegments.Where(s => s.IsChange))
            seg.State = ChangeState.Rejected;
        RebuildWorkingText();
        ReviewSegmentsChanged?.Invoke();
    }

    [RelayCommand]
    private void ApplyReview()
    {
        // Pending segments resolve to AddedText, so WorkingText already reflects
        // the fully-revised output. Just tear down the review UI.
        IsReviewing = false;
        ReviewSegments = [];
    }

    private void RebuildWorkingText()
    {
        WorkingText = string.Concat(ReviewSegments.Select(s => s.ResolvedText));
    }

    [RelayCommand]
    private void Undo()
    {
        if (_undoStack.Count > 0)
        {
            WorkingText = _undoStack.Pop();
            CanUndo = _undoStack.Count > 0;
            IsReviewing = false;
            HasResult = _undoStack.Count > 0;
            ReviewSegments = [];
        }
    }

    [RelayCommand]
    private async Task AcceptAsync()
    {
        // If still reviewing, finalize first
        if (IsReviewing)
            ApplyReview();

        var textToPaste = WorkingText;
        RequestClose?.Invoke();
        await Task.Delay(100);
        await _clipboardService.PasteTextAsync(textToPaste, PreviousForegroundWindow);
    }

    [RelayCommand]
    private void Dismiss()
    {
        _cts?.Cancel();
        RequestClose?.Invoke();
    }

    [RelayCommand]
    private void CopyResult()
    {
        if (!string.IsNullOrEmpty(WorkingText))
        {
            System.Windows.Application.Current.Dispatcher.Invoke(() =>
            {
                try { System.Windows.Clipboard.SetText(WorkingText); } catch { }
            });
        }
    }
}
